### ISSUE TITLE
fail_on_error clause makes logic engine facts False that would otherwise be true.

### ISSUE THREAD

**Issue by StevenCTimm** [HEPCloud/decisionengine#318]
I had previously tested that the fail_on_error clause successfully made facts "false" that would otherwise be "error".
But I am now observing is that some facts that would otherwise be True end up False when wrapped by fail_on_error.

Initially in resource_request.jsonnet I had the following:

          "gcewithininstburnrate": "fail_on_error(financial_params.iloc[0].target_gce_vm_burn_rate>GCE_Burn_Rate.iloc[0].BurnRate)",
          "gceabovebalance": "fail_on_error( financial_params.iloc[0].target_gce_balance<GCE_Billing_Info.iloc[0].Balance)",

For record, financial_params.iloc[0].target_gce_balance is -20000
GCE_Billing_Info.iloc.[0].Balance is -507
GCE_Burn_Rate.iloc.[0].BurnRate is 0.01
financial_params.iloc.[0].target_gce_vm_burn_rate is 9

Both of these facts wrapped by the fail_on_error evaluated to False.

I removed the fail_on_error wrapper and they evaluated to True as they should.

**Comment by knoepfel** [comment]
I have been able to reproduce the failure.  Investigating options.

**Comment by knoepfel** [comment]
The problem is understood.  I am exploring options--one of which is described here: https://stackoverflow.com/questions/66890900/replacing-python-call-ast-node-with-try-node

**Comment by knoepfel** [comment]
See PR #322.

**Comment by StevenCTimm** [comment]
PR looks promising do you have a rpm build artifact of it?  Could dig it out myself but appreciate if you could save me some time.  Steve

**Comment by knoepfel** [comment]
This should be the RPM: https://github.com/HEPCloud/decisionengine/suites/2393932846/artifacts/50894008

**Comment by StevenCTimm** [comment]
The patch does not seem to be working as designed.
Installed on fermicloud155, verified the patch is there.

2021-03-29T10:23:36-0500 - root - BooleanExpression - 4945 - MainThread - DEBUG - calling NamedFact::evaluate()
2021-03-29T10:23:36-0500 - root - datablock - 4945 - MainThread - ERROR - Did not get key in datablock __getitem__
2021-03-29T10:23:36-0500 - root - datablock - 4945 - MainThread - ERROR - No Key in datablock __getitem__

All the logic engine BooleanExpressions that are wrapped by fail_on_error now fail in this way

For reference:

          "awswithininstburnrate": "fail_on_error( financial_params.iloc[0].target_aws_vm_burn_rate>AWS_Burn_Rate.iloc[0].BurnRate)",
          "awswithinbillburnrate": "fail_on_error( financial_params.iloc[0].target_aws_bill_burn_rate>AWS_Billing_Rate[AWS_Billing_Rate['accountName']=='Fermilab'].iloc[0].costRatePerHourInLastSixHours )",
          "awsabovebalance": "fail_on_error( financial_params.iloc[0].target_aws_balance<AWS_Billing_Info[AWS_Billing_Info['AccountName']=='Fermilab'].iloc[0].Balance)",
          "gcewithininstburnrate": "fail_on_error( financial_params.iloc[0].target_gce_vm_burn_rate>GCE_Burn_Rate.iloc[0].BurnRate)",
          "gceabovebalance": "fail_on_error( financial_params.iloc[0].target_gce_balance<GCE_Billing_Info.iloc[0].Balance)",
          "fifenerscbelowlimit": "fail_on_error( Nersc_Allocation_Info[Nersc_Allocation_Info['name']=='fife'].iloc[0].usedAlloc<Nersc_Allocation_Info[Nersc_Allocation_Info['name']=='fife'].iloc[0].currentAlloc)",
          "uscmsnerscbelowlimit": "fail_on_error( Nersc_Allocation_Info[Nersc_Allocation_Info['name']=='uscms'].iloc[0].usedAlloc<Nersc_Allocation_Info[Nersc_Allocation_Info['name']=='uscms'].iloc[0].currentAlloc)"


I have verified that all of the quantities thus referenced in the above configuration file do exist so none of these
statements should be in an error condition at the moment.

**Comment by knoepfel** [comment]
Steve, I'm not positive that the logic-engine is to blame here--at least, not the `fail_on_error` component of it.  I suspect the exception is actually being raised in a (e.g.) downstream publisher.  I wouldn't know for sure, however, until I do some more debugging.  Any issues with me playing around with the `decisionengine` on fermicloud155?

**Comment by StevenCTimm** [comment]
go ahead, play with it.. note that the current configuration takes a while to start up, 10 minutes or so,.

**Comment by StevenCTimm** [comment]
all the goods are in /var/log/decisionengine/resource_request.log

**Comment by knoepfel** [comment]
Yep, it's my fault.  Modified the datablock code to print the missing key on fermicloud155:
```
2021-04-01T14:29:44-0500 - root - datablock - 15503 - MainThread - ERROR - Did not get key 'fail_on_error' in datablock __getitem__
```
Will look for a solution.

**Comment by StevenCTimm** [comment]
Followup--I believe that Kyle did indeed fix this issue but I have not yet verified it.. will attempt to add the
fail_on_error logic back to a 1.7 configuration to be sure it works.

**Comment by StevenCTimm** [comment]
I think so but I haven't had a chance to test it again recently.

Steve

________________________________

**Comment by StevenCTimm** [comment]
This appears to now be fixed.  (we have fail_on_error clauses in current DE config and they work)
